pkg/discovery/worker/aggregation/container_usage_data_aggregator.go

```diff
@@ -0,0 +1,84 @@
+package aggregation
+
+import (
+	"fmt"
+	"github.com/turbonomic/turbo-go-sdk/pkg/proto"
+	"math"
+)
+
+var (
+	avgUsageDataStrategy = "avgUsageData"
+	maxUsageDataStrategy = "maxUsageData"
+
+	DefaultContainerUsageDataAggStrategy = avgUsageDataStrategy
+
+	// Map from the configured utilization data aggregation strategy to utilization data aggregator
+	ContainerUsageDataAggregators = map[string]ContainerUsageDataAggregator{
+		avgUsageDataStrategy: &avgUsageDataAggregator{aggregationStrategy: "average usage data strategy"},
+		maxUsageDataStrategy: &maxUsageDataAggregator{aggregationStrategy: "max usage data strategy"},
+	}
+)
+
+// ContainerUsageDataAggregator interface represents a type of container usage data aggregator
+type ContainerUsageDataAggregator interface {
+	// AggregationStrategy returns aggregation strategy of this data aggregator
+	AggregationStrategy() string
+	// Aggregate aggregates commodities usage data based on the given list of commodity DTOs of a commodity type and
+	// aggregation strategy, and returns aggregated capacity, used and peak values.
+	Aggregate(containerCommodities []*proto.CommodityDTO) (float64, float64, float64, error)
+}
+
+// ---------------- Average usage data aggregation strategy ----------------
+type avgUsageDataAggregator struct {
+	aggregationStrategy string
+}
+
+func (avgUsageDataAggregator *avgUsageDataAggregator) AggregationStrategy() string {
+	return avgUsageDataAggregator.aggregationStrategy
+}
+
+func (avgUsageDataAggregator *avgUsageDataAggregator) Aggregate(commodities []*proto.CommodityDTO) (float64, float64, float64, error) {
+	if len(commodities) == 0 {
+		err := fmt.Errorf("error to aggregate commodities using %s : commodities list is empty",
+			avgUsageDataAggregator.AggregationStrategy())
+		return 0.0, 0.0, 0.0, err
+	}
+	capacitySum := 0.0
+	usedSum := 0.0
+	peakSum := 0.0
+	for _, commodity := range commodities {
+		capacitySum += *commodity.Capacity
```

Comment:
I think one or more of these values could be nil. Commodity could be created with only used value. Nil check is necessary here.

Reply:
And at all similar places.

Reply:
The commodities here are commodities sold which are supposed to have capacity set. But you are right, commodity could be created with only used value like commodities bought. Still it makes sense and safer to add nil check here.